File: ECommerce.Client/src/services/customerDashboardService.ts
import axios from 'axios';

const API_BASE_URL = 'http://localhost:5214/api';

export interface DashboardStats {
  totalOrders: number;
  totalSpent: number;
  pendingOrders: number;
}

export interface Order {
    orderId: number;
    orderDate: string;
    totalAmount: number;
    status: string;
    itemCount: number;
}

export interface CustomerProfile {
  customerId: number;
  fullName: string;
  email: string;
  joinDate: string;
  totalOrders: number;
  avatar?: string;
}

export interface DashboardData {
    stats: DashboardStats;
    recentOrders: Order[];
    profile: CustomerProfile;
}

class CustomerDashboardService {
    private getCustomerId(): number | null {
        const user = localStorage.getItem('user');
        if (user) {
            try {
                const parsedUser = JSON.parse(user);
                return parsedUser.id || null;
            } catch {
                return null;
            }
        }
        return null;
    }

    async getDashboardData(): Promise<DashboardData> {
        try {
            const customerId = this.getCustomerId();
            if (!customerId) {
                throw new Error('Customer ID not found');
            }

            const response = await axios.get(`${API_BASE_URL}/CustomerDashboard/${customerId}/dashboard`);
            return response.data;
        } catch (error) {
            console.error('Error fetching dashboard data:', error);
            throw error;
        }
    }

    async getStats(): Promise<DashboardStats> {
        try {
            const customerId = this.getCustomerId();
            if (!customerId) {
                throw new Error('Customer ID not found');
            }

            const response = await axios.get(`${API_BASE_URL}/CustomerDashboard/${customerId}/stats`);
            return response.data;
        } catch (error) {
            console.error('Error fetching stats:', error);
            throw error;
        }
    }

    async getRecentOrders(limit: number = 5): Promise<Order[]> {
        try {
            const customerId = this.getCustomerId();
            if (!customerId) {
                throw new Error('Customer ID not found');
            }

            const response = await axios.get(`${API_BASE_URL}/CustomerDashboard/${customerId}/recent-orders?limit=${limit}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching recent orders:', error);
            throw error;
        }
    }

    async getProfile(): Promise<CustomerProfile> {
        try {
            const customerId = this.getCustomerId();
            if (!customerId) {
                throw new Error('Customer ID not found');
            }

            const response = await axios.get(`${API_BASE_URL}/CustomerDashboard/${customerId}/profile`);
            return response.data;
        } catch (error) {
            console.error('Error fetching profile:', error);
            throw error;
        }
    }
}

export const customerDashboardService = new CustomerDashboardService();

File: ECommerce.Shared/Services/CustomerDashboardService.cs
using ECommerce.Shared.Models;
using ECommerce.Shared.DTOs;
using ECommerce.Shared.IServices;
using Microsoft.EntityFrameworkCore;

namespace ECommerce.Shared.Services;

public class CustomerDashboardService : ICustomerDashboardService
{
    private readonly KahreedoContext _context;

    public CustomerDashboardService(KahreedoContext context)
    {
        _context = context;
    }

    public async Task<CustomerDashboardDto> GetCustomerDashboardAsync(int customerId)
    {
        var customer = await _context.Customers
            .Include(c => c.Orders)
            .ThenInclude(o => o.OrderDetails)
            .FirstOrDefaultAsync(c => c.CustomerId == customerId);

        if (customer == null)
        {
            throw new Exception("Customer not found");
        }

        var completedOrders = customer.Orders.Where(o => o.IsCompleted == true).ToList();
        var pendingOrders = customer.Orders.Where(o => o.IsCompleted == false).ToList();

        var stats = new CustomerDashboardStatsDto
        {
            TotalOrders = completedOrders.Count,
            TotalSpent = completedOrders.Sum(o => o.TotalAmount ?? 0),
            PendingOrders = pendingOrders.Count
        };

        var profile = new CustomerProfileDto
        {
            CustomerId = customer.CustomerId,
            FullName = $"{customer.FirstName} {customer.LastName}",
            Email = customer.Email ?? "",
            JoinDate = customer.Created?.ToString("yyyy-MM-dd") ?? DateTime.Now.ToString("yyyy-MM-dd"),
            TotalOrders = completedOrders.Count,
            Avatar = customer.Picture
        };

        var recentOrders = customer.Orders
            .OrderByDescending(o => o.OrderDate)
            .Take(5)
            .Select(o => new RecentOrderDto
            {
                OrderId = o.OrderId,
                OrderDate = o.OrderDate?.ToString("yyyy-MM-dd") ?? "",
                TotalAmount = o.TotalAmount ?? 0,
                Status = GetOrderStatus(o),
                ItemCount = o.OrderDetails.Count
            })
            .ToList();

        return new CustomerDashboardDto
        {
            Stats = stats,
            Profile = profile,
            RecentOrders = recentOrders
        };
    }

    public async Task<CustomerDashboardStatsDto> GetCustomerStatsAsync(int customerId)
    {
        var customer = await _context.Customers
            .Include(c => c.Orders)
            .FirstOrDefaultAsync(c => c.CustomerId == customerId);

        if (customer == null)
        {
            throw new Exception("Customer not found");
        }

        var completedOrders = customer.Orders.Where(o => o.IsCompleted == true).ToList();
        var pendingOrders = customer.Orders.Where(o => o.IsCompleted == false).ToList();

        return new CustomerDashboardStatsDto
        {
            TotalOrders = completedOrders.Count,
            TotalSpent = completedOrders.Sum(o => o.TotalAmount ?? 0),
            PendingOrders = pendingOrders.Count
        };
    }

    public async Task<List<RecentOrderDto>> GetRecentOrdersAsync(int customerId, int limit = 5)
    {
        var orders = await _context.Orders
            .Where(o => o.CustomerId == customerId)
            .Include(o => o.OrderDetails)
            .OrderByDescending(o => o.OrderDate)
            .Take(limit)
            .Select(o => new RecentOrderDto
            {
                OrderId = o.OrderId,
                OrderDate = o.OrderDate!.Value.ToString("yyyy-MM-dd"),
                TotalAmount = o.TotalAmount ?? 0,
                Status = GetOrderStatus(o),
                ItemCount = o.OrderDetails.Sum(od => od.Quantity ?? 0)
            })
            .ToListAsync();

        return orders;
    }

    public async Task<CustomerProfileDto> GetCustomerProfileAsync(int customerId)
    {
        var customer = await _context.Customers
            .Include(c => c.Orders)
            .FirstOrDefaultAsync(c => c.CustomerId == customerId);

        if (customer == null)
        {
            throw new Exception("Customer not found");
        }

        var completedOrders = customer.Orders.Where(o => o.IsCompleted == true).Count();

        return new CustomerProfileDto
        {
            CustomerId = customer.CustomerId,
            FullName = $"{customer.FirstName} {customer.LastName}",
            Email = customer.Email ?? "",
            JoinDate = customer.Created?.ToString("yyyy-MM-dd") ?? "",
            TotalOrders = completedOrders,
            Avatar = customer.Picture
        };
    }

    private static string GetOrderStatus(Order order)
    {
        if (order.CancelOrder == true)
            return "Cancelled";
        
        if (order.Deliver == true)
            return "Delivered";
        
        if (order.Shipped == true)
            return "Shipped";
        
        if (order.Dispatched == true)
            return "Dispatched";
        
        if (order.IsCompleted == true)
            return "Completed";
        
        return "Processing";
    }
}
